python/ray/tune/experiment/trial.py

```diff
@@ -844,19 +867,17 @@ def on_restore(self):
     def should_recover(self):
         """Returns whether the trial qualifies for retrying.
 
-        This is if the trial has not failed more than max_failures. Note this
-        may return true even when there is no checkpoint, either because
+        `num_failures` should represent the number of times the trial has
+        failed *up to the moment this method is called.* If we've failed
+        5 times and `max_failures=5`, then we should recover, since
+        we only pass the limit on the 6th failure.
+
+        Note this may return true even when there is no checkpoint, either because
         `self.checkpoint_freq` is `0` or because the trial failed before
         a checkpoint has been made.
         """
         return (
-            self.run_metadata.num_failures < self.max_failures
-            or self.max_failures < 0
-            or (
-                self.run_metadata.num_failures == self.max_failures
-                and self.temporary_state.num_restore_failures
-                < int(os.environ.get("TUNE_RESTORE_RETRY_NUM", 0))
-            )
```

Comment:
I believe this condition is not needed anymore.

`TUNE_RESTORE_RETRY_NUM` configures how many attempts we try to restore before it counts as a real error.

The behavior with this condition removed makes sense to me:

If I'm at `num_failures==max_failures`, then I should try up to `TUNE_RESTORE_RETRY_NUM` times to restore. If all of those attempts fail, then we'll increment so that `num_failures > max_failures`, and the run will not try to recover anymore.